src/vidpf.rs

```diff
@@ -337,6 +338,41 @@ impl<W: VidpfValue, const NONCE_SIZE: usize> Vidpf<W, NONCE_SIZE> {
 
         proof
     }
+
+    /// Evaluates multiple user reports.
+    pub fn multi_eval(
+        &self,
+        reports: &Vec<ClientReport<W>>,
+        max_length: usize,
+    ) -> Result<(), VidpfError> {
+        let gen_prefixes = |n| -> Vec<BitVec> {
+            (0..(1usize << n))
+                .map(|i| i.view_bits::<Lsb0>()[..n].to_bitvec())
+                .collect()
+        };
+
+        for k in 1..max_length {
+            let hh_k = gen_prefixes(k);
+            for prefix in hh_k {
+                for ClientReport { key, public, nonce } in reports {
+                    let mut state = VidpfEvalState::init_from_key(key);
+                    // Creates a tree with root equal to the first state.
+                    let mut tree = BinaryTree::new(state.clone());
+
+                    let input = VidpfInput::from(prefix.clone());
+                    let nonce_array = &nonce[..NONCE_SIZE].try_into().unwrap(); // todo: make this conversion reliable
+
+                    (state, _) =
+                        self.eval_next(key.id, public, &input, k - 1, &state, nonce_array)?;
+
+                    // State is cached in the path indicated by the prefix.
+                    tree.insert(&prefix, state);
```

Comment:
Note that this is only evaluating the VIDPF at the first level, because it starts afresh with the initial state each time. The `BinaryTree` data structure only has one state inserted, and it isn't read from. To get the benefit of caching, we'd want to create a tree per client report, and use it to look up previously computed states while evaluating the same report at multiple prefixes. (and at multiple levels, though the multiple levels wouldn't all happen in one invocation in a realistic distributed trust setting)

Reply:
agree, I will let the high-level protocol to use the binary tree.